### PR TITLE
8311653: Modify -XshowSettings launcher behavior

### PR DESCRIPTION
**Pull request by coffeys**
Modify the -XshowSettings launcher option to print non-verbose settings details of all components by default. Verbose settings details will be printed via the -XshowSettings:all option. Modify the -XshowSettings option to reject bad values passed to it. Print an error message in such scenarios and abort the VM launch.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8312593](https://bugs.openjdk.org/browse/JDK-8312593) to be approved

### Issues
 * [JDK-8311653](https://bugs.openjdk.org/browse/JDK-8311653): Modify -XshowSettings launcher behavior (**Enhancement** - P4)
 * [JDK-8312593](https://bugs.openjdk.org/browse/JDK-8312593): Modify -XshowSettings launcher behavior (**CSR**)

### Reviewers
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15001/head:pull/15001` \
`$ git checkout pull/15001`

Update a local copy of the PR: \
`$ git checkout pull/15001` \
`$ git pull https://git.openjdk.org/jdk.git pull/15001/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15001`

View PR using the GUI difftool: \
`$ git pr show -t 15001`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15001.diff">https://git.openjdk.org/jdk/pull/15001.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15001#issuecomment-1648206622)

### CODE REVIEW

File: src/java.base/share/classes/sun/launcher/resources/launcher.properties [hunk 1]
```diff
@@ -226,6 +226,19 @@ The following options are macOS specific:\n\
 \    -Xdock:icon=<path to icon file>\n\
 \                      override default icon displayed in dock\n\n
 
+java.launcher.bad.option=\
+\n\
+Unrecognized showSettings option: {0}.\n\
+    Valid values are \"all\", \"locale\", \"properties\",\"security\",\
+    \"system\"(Linux only), \"vm\".\n\
```
Comment (mlchung) [inline src/java.base/share/classes/sun/launcher/resources/launcher.properties:233]: Nit:  space after commas missing.
```suggestion
    Valid values are \"all\", \"locale\", \"properties\", \"security\", \
    \"system\"(Linux only), \"vm\".\n\
```

File: src/java.base/share/classes/sun/launcher/LauncherHelper.java [hunk 2]
```diff
@@ -335,6 +317,11 @@ private static void printLocales() {
     }
 
     private static void printSystemMetrics() {
+        // only Linux supported
+        if (!OperatingSystem.isLinux()) {
+            abort(null, "java.launcher.bad.option");
```
Comment (mlchung) [inline src/java.base/share/classes/sun/launcher/LauncherHelper.java:322]:
```suggestion
            abort(null, "java.launcher.bad.option", "system");
```

maybe the `isLinux` check can be moved to `showSettings` and fall through to abort with the unmatched case.

File: src/java.base/share/classes/sun/launcher/LauncherHelper.java [hunk 3]
```diff
@@ -132,10 +123,13 @@ private static class ResourceBundleHolder {
...
-     * by default -XshowSettings is equivalent to -XshowSettings:all,
-     * Specific information may be gotten by using suboptions with possible
-     * values vm, properties and locale.
+     * A method called by the launcher to print out the standard settings.
+     * -XshowSettings prints details of all supported components in non-verbose
+     * mode. -XshowSettings:all prints all settings in verbose mode.
+     * Specific settings information may be obtained by using suboptions.
+     *
+     * suboption values include "all", "locale", "properties", "security",
+     * "system"(Linux only) and "vm". Help message printed for bad value.
```
Comment (RogerRiggs) [inline src/java.base/share/classes/sun/launcher/LauncherHelper.java:132]:
```suggestion
     * Suboption values include "all", "locale", "properties", "security",
     * "system" (Linux only) and "vm". A help message is printed for an unknown suboption value.
```

File: src/java.base/share/classes/sun/launcher/LauncherHelper.java [hunk 4]
```diff
@@ -159,43 +153,34 @@ static void showSettings(boolean printToStderr, String optionFlag,
...
-                    printSystemMetrics();
-                }
-                break;
+            case "all" -> printAllSettings(true, initialHeapSize, maxHeapSize, stackSize);
+            case "locale" -> printLocale(true);
+            case "properties" -> printProperties();
+            case "security" -> SecuritySettings.printSecuritySettings(opts, ostream, true);
+            case "system" -> printSystemMetrics();
+            case "vm" -> printVmSettings(initialHeapSize, maxHeapSize, stackSize);
+            case "" -> printAllSettings(false, initialHeapSize, maxHeapSize, stackSize);
```
Comment (RogerRiggs) [inline src/java.base/share/classes/sun/launcher/LauncherHelper.java:164]: There's mix in the arguments of where the "verbose" option is included.
Can they be consistent?  (For example put the verbose option as the last argument).

File: src/java.base/share/classes/sun/launcher/LauncherHelper.java [hunk 5]
```diff
@@ -159,43 +153,34 @@ static void showSettings(boolean printToStderr, String optionFlag,
...
+            default -> abort(null, "java.launcher.bad.option", optStr);
         }
     }
 
     /*
-     * prints the main vm settings subopt/section
+     * prints all available settings. Verbose option.
```
Comment (RogerRiggs) [inline src/java.base/share/classes/sun/launcher/LauncherHelper.java:170]: Don't take too many shortcuts in writing the description. Complete sentences are nice.

**Comment by coffeys** [comment]
Thanks for the feedback to date. I think I've captured all of it in recent update.

main changes: 
I've tightened up validation of input values to -XshowSettings and improved testing in this area. 
Updated java -X output to highlight difference between -XshowSettings and -XshowSettings:all

File: src/java.base/share/classes/sun/launcher/LauncherHelper.java [hunk 6]
```diff
@@ -158,44 +157,78 @@ static void showSettings(boolean printToStderr, String optionFlag,
...
+        if (optionFlag.equals("-XshowSettings")) {
+            return Option.EMPTY;
+        }
+
+        if (optionFlag.equals("-XshowSetings:")) {
+            abort(null, "java.launcher.bad.option", ":");
+        }
+
+        // case-sensitive check of input flag
+        List<String> validOpts = Arrays.stream(Option.values())
```
Comment (mlchung) [inline src/java.base/share/classes/sun/launcher/LauncherHelper.java:191]: The logic can be simplified by building a map of option name to `Option`.

```
        Map<String, Option> validOpts = Arrays.stream(Option.values())
                .filter(o -> !o.equals(Option.EMPTY)) // non-valid option
                .collect(Collectors.toMap(o -> o.name()
                                                .toLowerCase(Locale.ROOT)
                                                .replace("_", ":"), Function.identity()));

        String optStr = optionFlag.substring("-XshowSettings:".length());
        Option component = validOpts.get(optStr);
        if (component == null) {
            abort(null, "java.launcher.bad.option", optStr);
        }
        return component;
        
```

**Review by mlchung — APPROVED** [review]
Looks good.

Regarding the CSR, the specification change says:

> Bad options such as -XshowSettings:bad will no longer print any settings information. Instead a warning and help message is printed and the VM launch is aborted.

This should be updated as it will abort with an error message if the option is invalid.  No warning and help message is printed.

**Review by RogerRiggs — APPROVED** [review]
Looks good

**Comment by coffeys** [comment]
Thanks Mandy. CSR updated.

/integrate

**Pull request closed without merging**